### PR TITLE
Add Wasserstein distance penalty to discourage leakage of information into salient space

### PR DESCRIPTION
**Pull request by ethanweinberger**
This PR adds an additional penalty to the contrastiveVI/totalContrastiveVI modules that discourages shared variations from being captured in the contrastiveVI salient latent space. The high-level idea behind this penalty is to minimize the Wasserstein distance between the salient variable variational posteriors for background samples and an uninformative delta distribution centered at 0. As part of this PR, I also did some cleanup work in our model code. Specifically, this PR:

1. Adds the aforementioned penalty to our contrastiveVI/totalContrastiveVI module files with a new parameter (`wasserstein_penalty`) controlling the strength of the regularization.
2. Removes some old code implementing other regularizations that we found to be ineffective.
3. Updates our code for getting normalized expression values to be compatible with scvi-tools 0.16.
4. Modifies our contrastiveVI/totalContrastiveVI loss functions to follow a KL warmup schedule (now recommended by scvi-tools).

### CODE REVIEW

File: contrastive_vi/module/contrastive_vi.py [hunk 1]
```diff
@@ -578,60 +551,32 @@ def loss(
...
-        kl_z = background_losses["kl_z"] + target_losses["kl_z"]
-        kl_s = target_losses["kl_s"]
-        kl_library = background_losses["kl_library"] + target_losses["kl_library"]
-
-        loss = (
-            torch.sum(recon_loss)
-            + torch.sum(kl_z)
-            + torch.sum(kl_s)
-            + torch.sum(kl_library)
+        reconst_loss = background_losses["recon_loss"] + target_losses["recon_loss"]
```
Comment (ethanweinberger) [inline contrastive_vi/module/contrastive_vi.py:554]: This block of changes:

1. Adds our new Wasserstein penalty to the loss.
2. Modifies our loss function to incorporate KL warmup
3. Renames some variables to match the newest version of the scvi-tools example model (apologies for the noise with this).

File: contrastive_vi/model/total_contrastive_vi.py [hunk 2]
```diff
@@ -369,6 +371,7 @@ def get_normalized_expression(
         Return type is ``pd.DataFrame`` unless ``return_numpy`` is True.
         """
         adata = self._validate_anndata(adata)
+        adata_manager = self.get_anndata_manager(adata)
```
Comment (ethanweinberger) [inline contrastive_vi/model/total_contrastive_vi.py:374]: This + the remaining changes in this file simply make our normalized expression function compatible with scvi-tools 0.16 (currently it errors out).

File: contrastive_vi/module/total_contrastive_vi.py [hunk 3]
```diff
@@ -686,14 +692,15 @@ def loss(
...
+
+        wasserstein_loss = (
+            torch.norm(inference_outputs["background"]["qs_m"], dim=-1)**2
+            + torch.norm(torch.sqrt(inference_outputs["background"]["qs_v"]), dim=-1)**2
         )
 
-        kl_z = background_losses["kl_z"] + target_losses["kl_z"]
-        kl_s = target_losses["kl_s"]
-        kl_library_gene = (
+        kl_div_z = background_losses["kl_z"] + target_losses["kl_z"]
```
Comment (ethanweinberger) [inline contrastive_vi/module/total_contrastive_vi.py:701]: Similar to contrastiveVI, I modified the names and the order of the loss terms to follow that of the scvi-tools totalVI implementation (https://github.com/scverse/scvi-tools/blob/master/scvi/module/_totalvae.py)

File: contrastive_vi/module/contrastive_vi.py [hunk 4]
```diff
@@ -578,60 +551,37 @@ def loss(
...
+        kl_divergence_z = background_losses["kl_z"] + target_losses["kl_z"]
+        kl_divergence_s = target_losses["kl_s"]
+        kl_divergence_l = background_losses["kl_library"] + target_losses["kl_library"]
+
+        wasserstein_loss = (
```
Comment (chris522229197) [inline contrastive_vi/module/contrastive_vi.py:559]: Just to make sure. This is the squared Wasserstein's distance, right?

File: contrastive_vi/module/total_contrastive_vi.py [hunk 5]
```diff
@@ -686,14 +692,15 @@ def loss(
             background_losses["reconst_loss_protein"]
             + target_losses["reconst_loss_protein"]
         )
-        reconst_losses = dict(
-            reconst_loss_gene=reconst_loss_gene,
-            reconst_loss_protein=reconst_loss_protein,
+
+        wasserstein_loss = (
+            torch.norm(inference_outputs["background"]["qs_m"], dim=-1)**2
+            + torch.norm(torch.sqrt(inference_outputs["background"]["qs_v"]), dim=-1)**2
```
Comment (chris522229197) [inline contrastive_vi/module/total_contrastive_vi.py:698]: Again, this can be computed more efficiently as the sum of the variance terms.

File: contrastive_vi/module/total_contrastive_vi.py [hunk 6]
```diff
@@ -703,20 +710,30 @@ def loss(
...
+            reconst_loss_gene=reconst_loss_gene,
+            reconst_loss_protein=reconst_loss_protein,
+        )
         kl_local = dict(
-            kl_z=kl_z,
-            kl_s=kl_s,
-            kl_library_gene=kl_library_gene,
+            kl_div_z=kl_div_z,
+            kl_div_l_gene=kl_div_l_gene,
             kl_div_back_pro=kl_div_back_pro,
```
Comment (chris522229197) [inline contrastive_vi/module/total_contrastive_vi.py:727]: Why is `kl_div_s` ignored here?

File: contrastive_vi/module/total_contrastive_vi.py [hunk 7]
```diff
@@ -703,20 +710,30 @@ def loss(
...
         )
         kl_global = torch.tensor(0.0)
-        return LossRecorder(loss, reconst_losses, kl_local, kl_global)
+        return LossRecorder(
+            loss,
+            reconst_losses,
+            kl_local,
+            kl_global=kl_global,
+            wasserstein_loss=torch.sum(wasserstein_loss)
```
Comment (chris522229197) [inline contrastive_vi/module/total_contrastive_vi.py:735]: Again, maybe the mean should be taken here to be consistent with the overall loss.

File: contrastive_vi/module/contrastive_vi.py (hunk 4 above)
Comment (ethanweinberger) [reply contrastive_vi/module/contrastive_vi.py:559]: Yes, that's correct!

File: contrastive_vi/module/contrastive_vi.py [hunk 8]
```diff
@@ -578,60 +551,37 @@ def loss(
...
+            kl_divergence_s=kl_divergence_s
         )
         kl_global = torch.tensor(0.0)
-        return LossRecorder(loss, recon_loss, kl_local, kl_global)
+        return LossRecorder(
+            loss,
+            reconst_loss,
+            kl_local,
+            kl_global,
+            wasserstein_loss=torch.sum(wasserstein_loss)
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply contrastive_vi/module/contrastive_vi.py:583]: AFAICT, the `LossRecorder` internally sums any terms that aren't already a scalar before logging them. I.e., `loss` remains untouched (since we took the mean previously), but `recont_loss`, `kl_local`, and `kl_global`, are summed before being logged.

`LossRecorder` requires that any additional arguments for logging have already been reduced to scalars, so for consistency with the other individual components of the loss I took the sum of `wasserstein_loss`.

File: contrastive_vi/module/total_contrastive_vi.py (hunk 7 above)
Comment (ethanweinberger) [reply contrastive_vi/module/total_contrastive_vi.py:735]: See previous comment.

File: contrastive_vi/module/total_contrastive_vi.py (hunk 6 above)
Comment (ethanweinberger) [reply contrastive_vi/module/total_contrastive_vi.py:727]: Good catch! Fixed.

File: contrastive_vi/module/total_contrastive_vi.py (hunk 5 above)
Comment (ethanweinberger) [reply contrastive_vi/module/total_contrastive_vi.py:698]: Fixed

File: contrastive_vi/module/contrastive_vi.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply contrastive_vi/module/contrastive_vi.py:583]: I added a new comment to the code clarifying this.

File: contrastive_vi/module/contrastive_vi.py [hunk 9]
```diff
@@ -578,60 +551,37 @@ def loss(
...
-            + torch.sum(kl_s)
-            + torch.sum(kl_library)
+        reconst_loss = background_losses["recon_loss"] + target_losses["recon_loss"]
+        kl_divergence_z = background_losses["kl_z"] + target_losses["kl_z"]
+        kl_divergence_s = target_losses["kl_s"]
+        kl_divergence_l = background_losses["kl_library"] + target_losses["kl_library"]
+
+        wasserstein_loss = (
+            torch.norm(inference_outputs["background"]["qs_m"], dim=-1)**2
+            + torch.norm(torch.sqrt(inference_outputs["background"]["qs_v"]), dim=-1)**2
```
[the remark this answers is not in the document]
Comment (ethanweinberger) [reply contrastive_vi/module/contrastive_vi.py:561]: Fixed.

**Review by chris522229197 — APPROVED** [review]
LGTM!
